File: web/lib/quickLinks.ts
"use client";

/**
 * Quick Links shared logic — URL sanitization/validation and the "open in
 * the user's real browser" action.
 *
 * This mirrors the BACKEND sanitizer rules (defense in depth: the server is
 * the source of truth, this module exists so the client rejects garbage
 * before it ever reaches `POST/PUT /api/quick-links` and so the panel can
 * give instant feedback). Keep this file free of React imports so it stays
 * trivially unit-testable and importable from both the panel component and
 * an e2e harness.
 *
 * Accept/reject summary for `sanitizeUrl`:
 *  - Accepts: `https://…` (kept), `http://…` (upgraded to `https://`), and
 *    bare hosts with no scheme that contain a dot and no whitespace (e.g.
 *    `example.com/path` → `https://example.com/path`).
 *  - Rejects: empty/whitespace-only input, ASCII control characters anywhere
 *    in the string (0x00–0x1F, 0x7F — stripped first, then the remainder
 *    must still be non-empty), protocol-relative URLs (`//host/...`), and
 *    any explicit non-https scheme — `javascript:`, `data:`, `file:`,
 *    `vbscript:`, `blob:`, `about:`, `mailto:`, `tel:`, etc. The final
 *    candidate must parse via the `URL` constructor and resolve to
 *    `protocol === "https:"` with a non-empty `hostname`.
 *  - On invalid input `sanitizeUrl` THROWS `InvalidUrlError` (does not
 *    return null) so callers can't accidentally treat an unsanitized string
 *    as safe by forgetting a null check.
 */

/** Thrown by {@link sanitizeUrl} when `raw` cannot be turned into a safe,
 * https-only, non-empty-host URL. */
export class InvalidUrlError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidUrlError";
  }
}

/** ASCII control characters (0x00–0x1F, 0x7F), matched anywhere in the string. */
const CONTROL_CHARS_RE = /[\x00-\x1F\x7F]/g;

/** Matches a leading URI scheme, e.g. `javascript:`, `mailto:`, `data:`. */
const HAS_SCHEME_RE = /^[a-zA-Z][a-zA-Z0-9+.-]*:/;

/**
 * Normalize `raw` into an `https://` candidate string, or throw
 * `InvalidUrlError`. Does not itself parse with `URL` — that final
 * validation happens in {@link sanitizeUrl}.
 */
function toHttpsCandidate(raw: string): string {
  const cleaned = raw.replace(CONTROL_CHARS_RE, "").trim();
  if (!cleaned) {
    throw new InvalidUrlError("URL is empty");
  }
  if (cleaned.startsWith("//")) {
    throw new InvalidUrlError(`Protocol-relative URLs are not allowed: ${raw}`);
  }
  if (/^https:\/\//i.test(cleaned)) {
    return cleaned;
  }
  if (/^http:\/\//i.test(cleaned)) {
    return `https://${cleaned.slice("http://".length)}`;
  }
  if (HAS_SCHEME_RE.test(cleaned)) {
    // javascript:, data:, file:, vbscript:, blob:, about:, mailto:, tel:, etc.
    throw new InvalidUrlError(`Unsupported URL scheme: ${raw}`);
  }
  // No scheme at all — only accept if it looks like a bare host: has a dot
  // and contains no whitespace (rules out things like "not a url").
  if (/\s/.test(cleaned) || !cleaned.includes(".")) {
    throw new InvalidUrlError(`Not a valid URL: ${raw}`);
  }
  return `https://${cleaned}`;
}

/**
 * Sanitize and validate a user-supplied Quick Link URL. See the module
 * doc-comment above for the full accept/reject matrix.
 *
 * @throws {InvalidUrlError} if `raw` cannot be resolved to a valid
 *   `https:` URL with a non-empty host.
 */
export function sanitizeUrl(raw: string): string {
  const candidate = toHttpsCandidate(raw);
  let parsed: URL;
  try {
    parsed = new URL(candidate);
  } catch {
    throw new InvalidUrlError(`Not a valid URL: ${raw}`);
  }
  if (parsed.protocol !== "https:" || !parsed.hostname) {
    throw new InvalidUrlError(`Only https URLs with a host are allowed: ${raw}`);
  }
  return parsed.toString();
}

/** True iff {@link sanitizeUrl} would accept `raw` without throwing. */
export function isValidUrl(raw: string): boolean {
  try {
    sanitizeUrl(raw);
    return true;
  } catch {
    return false;
  }
}

// --- custom icons -----------------------------------------------------------

/** Prefix every stored image icon must carry (mirrors the backend). */
const IMAGE_DATA_URI_PREFIX = "data:image/png;base64,";
/** Preset keys accepted by the backend (`^[a-z0-9_-]{1,32}$`). */
const PRESET_KEY_RE = /^[a-z0-9_-]{1,32}$/;
/** ~24 KB cap on a stored image data URI, matching `_MAX_ICON_DATA_LEN`. */
const MAX_ICON_DATA_LEN = 24_000;

/**
 * Client-side mirror of `backend/quick_links.sanitize_icon_value` (defense in
 * depth, same rationale as {@link sanitizeUrl}): the server is the source of
 * truth, but validating here lets the panel reject a too-big image before the
 * request and keeps the picker honest.
 *
 * - `kind === null` → no custom icon; any value is fine (ignored).
 * - `"preset"` → short `[a-z0-9_-]` key.
 * - `"image"`  → a `data:image/png;base64,…` URI within the size cap.
 */
export function isValidIconValue(kind: string | null, value: string | null): boolean {
  if (kind === null) return true;
  if (!value) return false;
  if (kind === "preset") return PRESET_KEY_RE.test(value);
  if (kind === "image") {
    return value.startsWith(IMAGE_DATA_URI_PREFIX) && value.length <= MAX_ICON_DATA_LEN;
  }
  return false;
}

/**
 * Downscale any picked image `File` to a square PNG **data URI** (default
 * 64x64), preserving aspect ratio and centering on a transparent canvas.
 *
 * This is the single normalization point for *both* icon-upload paths — the
 * web `<input type="file">` and the Electron `argus.pickIcon()` dialog — so the
 * value stored in SQLite is always a small `data:image/png` regardless of the
 * source format (png/jpg/webp/svg) or original dimensions. The result renders
 * under the existing `img-src 'self' data: blob:` CSP with no policy change.
 *
 * Rejects (rejects the promise) if the file isn't a decodable image or the
 * downscaled result would still exceed the stored-size cap.
 */
export function fileToPngDataUrl(file: File | Blob, size = 64): Promise<string> {
  return new Promise((resolve, reject) => {
    const objectUrl = URL.createObjectURL(file);
    const img = new Image();
    img.onload = () => {
      URL.revokeObjectURL(objectUrl);
      try {
        const canvas = document.createElement("canvas");
        canvas.width = size;
        canvas.height = size;
        const ctx = canvas.getContext("2d");
        if (!ctx) {
          reject(new Error("Canvas is not available"));
          return;
        }
        // "contain" fit: scale the longest side to `size`, center the rest.
        const scale = Math.min(size / img.width, size / img.height) || 1;
        const w = Math.max(1, Math.round(img.width * scale));
        const h = Math.max(1, Math.round(img.height * scale));
        ctx.drawImage(img, (size - w) / 2, (size - h) / 2, w, h);
        const dataUrl = canvas.toDataURL("image/png");
        if (!dataUrl.startsWith(IMAGE_DATA_URI_PREFIX) || dataUrl.length > MAX_ICON_DATA_LEN) {
          reject(new Error("Image is too large after downscaling"));
          return;
        }
        resolve(dataUrl);
      } catch (error) {
        reject(error instanceof Error ? error : new Error("Could not process image"));
      }
    };
    img.onerror = () => {
      URL.revokeObjectURL(objectUrl);
      reject(new Error("Could not load image"));
    };
    img.src = objectUrl;
  });
}

/**
 * Decode a `data:<mime>;base64,<body>` URI to a `Blob` without any network
 * request. Used for the Electron `argus.pickIcon()` result — `fetch()` on a
 * `data:` URL can be blocked by the loopback `connect-src` CSP, so we parse it
 * by hand instead. Throws on a non-base64 `data:` URI.
 */
function dataUrlToBlob(dataUrl: string): Blob {
  const match = /^data:([^;,]*)?(;base64)?,(.*)$/s.exec(dataUrl);
  if (!match) throw new Error("Not a data URI");
  const mime = match[1] || "application/octet-stream";
  const isBase64 = Boolean(match[2]);
  const body = match[3];
  const binary = isBase64 ? atob(body) : decodeURIComponent(body);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) bytes[i] = binary.charCodeAt(i);
  return new Blob([bytes], { type: mime });
}

/**
 * Turn a raw `data:` URI (e.g. the value the Electron main process returns from
 * `argus.pickIcon()`) into a downscaled PNG data URI via {@link fileToPngDataUrl}.
 */
export async function dataUrlToPngDataUrl(dataUrl: string, size = 64): Promise<string> {
  return fileToPngDataUrl(dataUrlToBlob(dataUrl), size);
}

/**
 * Open a Quick Link in the user's real browser. Under the Electron desktop
 * shell this goes through the `window.argus.openExternal` preload bridge
 * (see `desktop/preload.js` / `desktop/main.js`'s `shell:open` handler,
 * which itself only allows `https:`/`obsidian:`); in the plain web app it
 * falls back to `window.open` in a new tab.
 *
 * The URL is always sanitized first via {@link sanitizeUrl} so an invalid
 * or unsafe string can never reach `openExternal`/`window.open`. If
 * sanitization fails this function silently does nothing (it does not
 * throw) — callers that want to surface an error to the user should call
 * {@link isValidUrl} (or `sanitizeUrl` directly) beforehand.
 */
export function openExternalUrl(url: string): void {
  let safe: string;
  try {
    safe = sanitizeUrl(url);
  } catch {
    return;
  }
  const w = window as unknown as { argus?: { openExternal?: (u: string) => Promise<boolean> | boolean } };
  if (w.argus?.openExternal) {
    void w.argus.openExternal(safe);
  } else {
    window.open(safe, "_blank", "noopener,noreferrer");
  }
}

File: backend/quick_links.py
"""Quick Links: user-curated shortcut buttons shown on the dashboard.

Mirrors ``backend/flashcards.py``'s plain-``sqlite3``, plain-dict CRUD style.
The one thing that makes this module security-sensitive is
:func:`sanitize_url`: quick links round-trip through the dashboard as plain
``<a href>`` targets, so anything that isn't a real ``https://`` URL (a
``javascript:`` URI, a protocol-relative ``//evil.com``, a control-character
obfuscated scheme, ...) must be rejected before it ever reaches the database.
"""

from __future__ import annotations

import re
import sqlite3
from urllib.parse import urlparse

_MAX_LABEL_LEN = 80
_MAX_ICON_LEN = 8

# ASCII control characters (0x00-0x1F, 0x7F) stripped from anywhere in the
# raw URL before any other check runs. This defeats obfuscation like
# ``java\tscript:`` or ``java\nscript:``, which otherwise collapse into a
# working ``javascript:`` scheme once whitespace-trimmed by a naive check.
_CONTROL_CHARS_RE = re.compile(r"[\x00-\x1f\x7f]")

# Pseudo-schemes that must never be stored, regardless of case or of whether
# the rest of the string looks like a URL.
_DANGEROUS_SCHEMES = (
    "javascript:",
    "data:",
    "file:",
    "vbscript:",
    "blob:",
    "about:",
    "mailto:",
    "tel:",
)


class QuickLinksError(Exception):
    """Raised on invalid quick-link input (bad url/label/icon, or unknown id)."""


def sanitize_url(raw: str) -> str:
    """Normalize ``raw`` to a safe ``https://`` URL, or raise :class:`QuickLinksError`.

    - Strips leading/trailing whitespace and ASCII control characters
      anywhere in the string.
    - ``http://`` is upgraded to ``https://``; ``https://`` is kept as-is.
    - A schemeless value that looks like a host (contains a dot, no spaces)
      is treated as ``https://<value>``.
    - Dangerous pseudo-schemes (``javascript:``, ``data:``, ``file:``,
      ``vbscript:``, ``blob:``, ``about:``, ``mailto:``, ``tel:``) and
      protocol-relative (``//...``) URLs are always rejected, as is any
      other scheme that isn't ``http``/``https``.
    - The final result must parse (via :func:`urllib.parse.urlparse`) with
      ``scheme == "https"`` and a non-empty ``netloc``.
    """
    if raw is None:
        raise QuickLinksError("url is required")
    cleaned = _CONTROL_CHARS_RE.sub("", raw).strip()
    if not cleaned:
        raise QuickLinksError("url is required")

    # Only the scheme is ever lowercased for comparison — never the whole
    # URL, since path/query/fragment can be case-sensitive.
    lowered = cleaned.lower()

    if lowered.startswith("//"):
        raise QuickLinksError("protocol-relative URLs are not allowed")

    for scheme in _DANGEROUS_SCHEMES:
        if lowered.startswith(scheme):
            raise QuickLinksError(f"scheme not allowed: {scheme[:-1]}")

    if "://" in cleaned:
        scheme, _, rest = cleaned.partition("://")
        scheme_lower = scheme.lower()
        if scheme_lower in ("http", "https"):
            normalized = f"https://{rest}"
        else:
            raise QuickLinksError(f"scheme not allowed: {scheme_lower or '(empty)'}")
    else:
        # No scheme at all: only accept it if it plausibly names a host.
        if " " in cleaned or "." not in cleaned:
            raise QuickLinksError("url must be a valid https:// URL")
        normalized = f"https://{cleaned}"

    parsed = urlparse(normalized)
    if parsed.scheme != "https" or not parsed.netloc:
        raise QuickLinksError("url must be a valid https:// URL with a host")
    return normalized


def sanitize_label(raw: str) -> str:
    """Trim, collapse internal whitespace, and cap the length of a label."""
    if raw is None:
        raise QuickLinksError("label is required")
    collapsed = " ".join(raw.split())
    if not collapsed:
        raise QuickLinksError("label is required")
    return collapsed[:_MAX_LABEL_LEN]


def sanitize_icon(raw: str | None) -> str | None:
    """Trim an icon glyph; empty/None becomes ``None``."""
    if raw is None:
        return None
    trimmed = raw.strip()
    return trimmed[:_MAX_ICON_LEN] or None


def _row_to_dict(row: sqlite3.Row) -> dict:
    return {
        "id": row["id"],
        "created_at": row["created_at"],
        "label": row["label"],
        "url": row["url"],
        "icon": row["icon"],
        "sort_order": row["sort_order"],
    }


def _get_link_row(conn: sqlite3.Connection, link_id: int) -> sqlite3.Row:
    row = conn.execute(
        "SELECT id, created_at, label, url, icon, sort_order FROM quick_links WHERE id = ?",
        (link_id,),
    ).fetchone()
    if row is None:
        raise QuickLinksError(f"quick link not found: {link_id}")
    return row


def list_links(conn: sqlite3.Connection) -> list[dict]:
    """All quick links, in display order."""
    rows = conn.execute(
        "SELECT id, created_at, label, url, icon, sort_order FROM quick_links"
        " ORDER BY sort_order, id"
    ).fetchall()
    return [_row_to_dict(row) for row in rows]


def create_link(conn: sqlite3.Connection, *, label: str, url: str, icon: str | None) -> dict:
    """Sanitize and insert a new quick link, appended to the end of the order."""
    clean_label = sanitize_label(label)
    clean_url = sanitize_url(url)
    clean_icon = sanitize_icon(icon)

    max_row = conn.execute("SELECT MAX(sort_order) AS max_order FROM quick_links").fetchone()
    next_order = (max_row["max_order"] or 0) + 1

    cursor = conn.execute(
        "INSERT INTO quick_links (label, url, icon, sort_order) VALUES (?, ?, ?, ?)",
        (clean_label, clean_url, clean_icon, next_order),
    )
    conn.commit()
    return _row_to_dict(_get_link_row(conn, int(cursor.lastrowid)))


def update_link(
    conn: sqlite3.Connection,
    link_id: int,
    *,
    label: str | None = None,
    url: str | None = None,
    icon: str | None = None,
    sort_order: int | None = None,
) -> dict:
    """Update only the provided fields (also used for drag-to-reorder via ``sort_order``).

    Raises :class:`QuickLinksError` if ``link_id`` doesn't exist.
    """
    _get_link_row(conn, link_id)  # raises QuickLinksError if missing

    fields: list[str] = []
    values: list[object] = []
    if label is not None:
        fields.append("label = ?")
        values.append(sanitize_label(label))
    if url is not None:
        fields.append("url = ?")
        values.append(sanitize_url(url))
    if icon is not None:
        fields.append("icon = ?")
        values.append(sanitize_icon(icon))
    if sort_order is not None:
        fields.append("sort_order = ?")
        values.append(sort_order)

    if fields:
        values.append(link_id)
        conn.execute(f"UPDATE quick_links SET {', '.join(fields)} WHERE id = ?", values)
        conn.commit()

    return _row_to_dict(_get_link_row(conn, link_id))


def delete_link(conn: sqlite3.Connection, link_id: int) -> None:
    """Delete a quick link. Raises :class:`QuickLinksError` if it doesn't exist."""
    _get_link_row(conn, link_id)  # raises QuickLinksError if missing
    conn.execute("DELETE FROM quick_links WHERE id = ?", (link_id,))
    conn.commit()
